src/test/java/io/vertx/core/http/HttpBandwidthLimitingTest.java

```diff
@@ -0,0 +1,259 @@
+  }
+
+  /**
+   * The throttling takes a while to kick in so the expected time cannot be strict especially
+   * for small data sizes in these tests.
+   *
+   * @param size
+   * @param rate
+   * @return
+   */
+  private long expectedTimeMillis(int size, int rate) {
+    return (long) (TimeUnit.MILLISECONDS.convert((size / rate), TimeUnit.SECONDS) * 0.5);
+  }
+
+  private void read(Buffer expected, HttpServer server, HttpClient client) {
+    client.request(HttpMethod.GET, server.actualPort(), DEFAULT_HTTP_HOST,"/buffer-read",
+      req -> {
+        req.result().send(resp -> resp.result().bodyHandler(r -> {
+          assertEquals(expected.getByteBuf(), r.getByteBuf());
+          testComplete();
+        }));
+      });
+  }
+
+  private void write(Buffer buffer, HttpServer server, HttpClient client) {
+    client.request(HttpMethod.POST, server.actualPort(), DEFAULT_HTTP_HOST, "/buffer-write",
+                   req -> req.result()
+                             .putHeader(HttpHeaderNames.CONTENT_LENGTH, String.valueOf(buffer.length()))
+                             .end(buffer)
+                   );
+  }
+
+  private void upload(HttpServer server, HttpClient client, File expected) {
+    Buffer b = vertx.fileSystem().readFileBlocking(expected.getAbsolutePath());
+    client.request(HttpMethod.PUT, server.actualPort(), DEFAULT_HTTP_HOST, "/upload-file", req -> {
+      req.result()
+         .putHeader(HttpHeaderNames.CONTENT_LENGTH, String.valueOf(expected.length()))
+         .putHeader(HttpHeaderNames.CONTENT_TYPE, "application/binary")
+         .end(b);
+    });
+  }
+
+  class Handlers {
+    public Handler<HttpServerRequest> bufferRead(Buffer expectedBuffer) {
+      return (req) -> {
+        req.response().setChunked(true);
+
+        int start = 0;
+        int size = expectedBuffer.length();
+        int chunkSize = OUTBOUND_LIMIT / 2;
+        while (size > 0) {
+          int len = Math.min(chunkSize, size);
+          req.response().write(expectedBuffer.getBuffer(start, start + len));
+          start += len;
+          size -= len;
+        }
+        req.response().end();
+      };
+    }
+
+    public Handler<HttpServerRequest> getFile(File expected) {
+      return req -> req.response().sendFile(expected.getAbsolutePath());
+    }
+
+    public Handler<HttpServerRequest> bufferWrite(Buffer expected) {
+      return req -> {
+        req.bodyHandler(buffer -> {
+          assertEquals(expected.getByteBuf(), buffer.getByteBuf());
+          testComplete();
+        });
+      };
+    }
+
+    public Handler<HttpServerRequest> uploadFile(File expected) {
+      return req -> {
+        req.endHandler((r) -> {
+          assertEquals(expected.length(), req.bytesRead());
+          testComplete();
+        });
+      };
+    }
+  }
+
+  static class Providers {
+    private static HttpServer http1Server(Vertx vertx, int inboundLimit, int outboundLimit) {
+      int openPort = findFreePort();
```

Comment:
you can listen on port 0 that will give you a randomly available port, then the port can be retrieved using `server.actualPort()`